### ISSUE TITLE
alternative install with docker

### ISSUE THREAD

**Issue by khassel** [Bee-Mar/mmpm#21]
Hi,

thank you for offering this nice module!

May someone is interested using a docker image instead of installing this stuff directly on his pi.

Did a quick setup [here](https://gitlab.com/khassel/mmpm), this could be a starting point ...

Let me know if this is of interest ...

**Comment by Bee-Mar** [comment]
@khassel This is definitely of interest! I was actually considering creating a docker image about a month ago, but never got around to it. I'll pull the image, give it a whirl, and see how it behaves. Once you think it's operating well enough, we can figure out how to link the two projects together (either combining it into this repo, or just leaving it as a link on the readme here...literally just spitballing as I type this)

Seriously, this is great, thank you!

**Comment by Bee-Mar** [comment]
At the moment, I'm converting the backend of things to use `sqlalchemy` with a proper SQLite database, to improve the performance. Which, really should have been done in the first place, but whaterver lol

**Comment by khassel** [comment]
Thanks 🙂

Test it ... 

I'm already thinking about 
* a setup where MagicMirror is also running as docker image (this works already, documentation in progress ...)
* may getting an enhancement (e.g. env var) disabling the `control center` tab which will not work with docker

**Comment by Bee-Mar** [comment]
@khassel I've added a dedicated wiki page containing basic information for you docker image, as well as link in the primary installation wiki page

**Comment by khassel** [comment]
Thanks!

Yesterday I added instructions to run the mmpm-docker-setup together with a mm-docker-setup.
There is also now a docker image for amd64-architecture (beside arm) for people running mm in serveronly mode on a linux machine.

@Bee-Mar I don't know if you looked into my Docker setup, you wrote your app an you know best how to deploy this stuff. So don't hesitate with improvements ...

**Comment by khassel** [comment]
Hi,

I have a question regarding the ports 7892/7893 for webSSHLocation.

Why is this proxied via nginx?

Would `webSSHLocation = 'http://localhost:7893';` in `app.components.ts` not be sufficient and you could omit port 7892?

Thanks,

Karsten.

**Comment by Bee-Mar** [comment]
using `http://localhost:7893' without the proxy would be pointing to the localhost of whatever device is being used to view the webpage, which would prevent that URL from being resolved properly

**Comment by khassel** [comment]
sorry for the noise, was in debug mode while testing this and forgot that angular works on the client side ...

**Comment by khassel** [comment]
one point left is

>    may getting an enhancement (e.g. env var) disabling the `control center` tab which will not work with docker

At the moment I'm doing an ugly hack to operate the `control center` tab out of the `app.component.html`:

```bash
    perl -pi.old -e 's|>\n|>|g' gui/src/app/app.component.html; \
    perl -pi -e 's|<mat-tab label="Control Center">.*</mat-tab>||g' gui/src/app/app.component.html; \
    perl -pi -e 's|>|>\n|g' gui/src/app/app.component.html; \
```
Would really appreciate getting a more solid solution here ...

**Comment by Bee-Mar** [comment]
@khassel not a problem at all! In addition to a total rework of the CLI, I'm looking to integrate some of the control center into the socket.io sockets being used by MagicMirror. As far as being able to manage the power of the raspberry pi from within the docker image, that might just be something that's best left as something that shouldn't be hacked into for security purposes. Or, left as a flag for the user to opt into within the docker-compose using an environment variable, if that functionality is really desirable

**Comment by khassel** [comment]
> and looking to integrate some of the control center into the socket.io sockets being used by MagicMirror. As far as being able to manage the power of the raspberry pi from within the docker image

With raspian stretch it was possible to map all things into the container needed for doing a restart/shutdown from inside the container. But don't ask for security doing such things.

With buster this is no longer possible (may there are other ugly hacks).

What you can always do is to write a service which must be installed on the host and the container triggers actions by e.g. writing in files the service is watching on. But this implies a service install on the host beside the docker stuff.

I didn't understand the idea using socket.io from MagicMirror for control center actions. The MagicMirror may also runs as container and therefore has (in such a case) no access to power management ...

**Comment by Bee-Mar** [comment]
ah, well, that's unfortunate the ability to power off or restart the raspberry pi isn't available anymore. Sorry, I was not thinking very clearly when I brought up the socket.io stuff, it's not actually relevant to the topic lol.

Also, once I have the upgrades done for the next major release of MMPM (hopefully in the next 2 weeks) I'm planning on helping you out with the docker image. I noticed there have been quite a few downloads of the docker image on docker hub. It's up to something like 6600 pulls!

**Comment by Bee-Mar** [comment]
@khassel Just a friendly heads up so I don't break your docker image. I'm still working on the major upgrades, which are getting quite close to being done. Once everything is merged into develop (probably a rebase in reality) ll let you know. I'll let it sit in develop while I work on writing some tests, and do some testing as well

**Comment by Bee-Mar** [comment]
@khassel so, I'm right on track, and the only thing remaining is the theme switching, but it's definitely at the point where you can give it a shot. The revamped version is on the develop branch. 

You can install it by running `make from-src` at  root of the repo

The one-liner is `git clone https://github.com/bee-mar/mmpm.git -b develop; cd mmpm; make from-src`

and of course, `$HOME/.local/bin` needs to be in your `$PATH`

**Comment by khassel** [comment]
@Bee-Mar the first version is running ... I was a bit surprised of the theme 😜

There were some minor issues, no problem to solve, but I mention these things because they may cause problems in normal installs.

New todo's before install (doing the install as root):
````
PATH=$PATH:/root/.local/bin
mkdir -p /root/.local/bin
mkdir -p /root/.config/mmpm
touch /root/.config/mmpm/mmpm-env.json
mkdir -p /root/.config/mmpm/log
````

I'm not happy with the new feature, having all the env vars in `mmpm-env.json`.
This file can now be changed via mmpm-editor, so I have to mount it to the host (otherwise changes are lost).
I see no benefit of editing this file via mmpm, because you have to restart mmpm to apply changes
(which is not possible in docker setup).

So the standard solution to use classic env variables would be much better (at least for the docker setup).
With this approach you would define these variables in the `docker-compose.yml`.

Or I don't mount the `mmpm-env.json` and create it before starting from corresponding env vars, 
but then it should not be editable in mmpm when running as docker container.

Last thing is the `MMPM_MAGICMIRROR_DOCKER_COMPOSE_FILE` var. The `docker-compose.yml` exists only on the host, not in the container. You could mount it in the container but it would be useless, because you need to install `docker` and `docker-compose` in the container only for starting/stopping (not talking about security implications allowing a container to use docker from the host). There is no easy way to implement this use case. One approach (only for restarting) could be to kill the main process of the container and restart it automatically in such a case, will do some tests in this direction ...

**Comment by Bee-Mar** [comment]
mounting `mmpm-env.json` on the host is the intention. If the user was changing settings through environment variables, the entire web server would need to be stopped, then restarted. But by the contents being in the JSON file, it can be reloaded immediately. All you would need to do is refresh the page. 

In the case of the docker version of MMPM, there is no need to set `MMPM_MAGICMIRROR_DOCKER_COMPOSE_FILE`, that should be left as an empty string.

**Comment by khassel** [comment]
> That setting is only if MMPM is not a docker image, but MagicMirror is

missed that use case ...

One thing I'm struggling with is 
![grafik](https://user-images.githubusercontent.com/25914086/87860691-bed56480-c93f-11ea-8122-8ab8c2b54a37.png)

Looking into code found `this.socket = io(`${this.mmpmEnvVars?.get("MMPM_MAGICMIRROR_URI")}/mmpm`, {reconnection: true});` so you try to reach `<mm-uri>/mmpm` for this. But this url doesn't work. Installed already mmpm as module with minimal config
````
		{
			module: "mmpm",
			position: "top_left"
		},
````

````
[2020-07-18 21:31:51.350] [LOG]    Starting module helper: mmpm
````

but with no success. Any idea what I missed?

**Comment by Bee-Mar** [comment]
yeah, sorry forgot to mention this lol

so, after installing the new version of mmpm, you need to create a module with `{ module: "mmpm" }`, then it also make sure the MagicMirror config.js ipWhitelist is exposing the host ip address of the MagicMirror

so `ipWhitelist: []` will make everything visible, and then you might also need to open the magicmirror in the browser, sometimes that's needed

**Comment by Bee-Mar** [comment]
you can leave out the position value if you want

**Comment by khassel** [comment]
now working, didn't change anything, just restarted both containers ...

Another thing was that `mmpm` was not in the module list, so I had to install it manually and add it to `External Packages`.

Concerning the `mmpm-env.json` I decided to mount the whole `/home/node/.config/mmpm` directory because this contains other stuff which should be saved on the host e.g. `mmpm-external-packages.json`

**Comment by Bee-Mar** [comment]
@khassel awesome! yeah, mounting that whole directory will make things much easier for the whole process. I'm going to lay out all this stuff as clear as possible in the wiki. Also, the styling is still changing, but the core functionality appears to be solid, so once the release is up on github then you should be good to set it all up on the docker image

**Comment by Bee-Mar** [comment]
also, what version of python3 did you use when setting this image up? locally, python3.7 or greater has been the sweet spot

**Comment by khassel** [comment]
I'm using `node:buster-slim` as base image and install python3 with `apt-get install -y python3 python3-pip`.
At the moment, this is Python 3.7.3

**Comment by Bee-Mar** [comment]
cool, that should be great

**Comment by Bee-Mar** [comment]
also, thank you a bunch for 1) deciding to create the docker image of this, and 2) maintaining it

**Comment by khassel** [comment]
Finally I did an automated build and tested the rpi-image.

The images using your `develop` branch are available as `karsten13/mmpm:develop`.

Meanwhile (already before the new mmpm release) I ended up with this build process:
- build gui (results can be used by arm and amd64)
- build arm and amd64 executables
- put things together in slim runtime-image

**Comment by Bee-Mar** [comment]
nice! that's a good test bed for the whole thing
i'll give it a shot later today also

**Comment by Bee-Mar** [comment]
@khassel also, I just realized I hadn't mentioned this, but the `node_helper.js` and `mmpm.js` files at the root of the repo should be added in a module folder,

assuming you've cloned `mmpm` into your home directory

so `mkdir -p MMPM_MAGICMIRROR_ROOT/modules/mmpm ; ~/mmpm/*js MMPM_MAGICMIRROR_ROOT/modules/mmpm`

that'll activate the hiding and toggling

**Comment by khassel** [comment]
> Another thing was that mmpm was not in the module list, so I had to install it manually and add it to External Packages.

thats what I already meant with the sentence above ...

With the new release `mmpm` needs to be installed as module in mm, but is not avaliable to choose from within the `mmpm` gui.

So I think it should be installable from within the gui or the manual steps you mentioned above should be done automatically (plus inserting in `config.js` ?).

**Comment by Bee-Mar** [comment]
gotcha. I'll definitely make the instructions clear for people on the wiki. Also, I'm going to have the CLI available on PyPI. It's up there right now, so you can actually just run `pip3 install mmpm` from terminal without having to clone the repo. I'm going to see how the GUI might get bundled into something, or maybe have it up on npm. Havent' quite figured that out just yet lol

**Comment by Bee-Mar** [comment]
@khassel i'm prepping to release everything shortly, and the installation method is almost certainly not going to work well for the docker image. I can setup a suitable `make` target for you so that whenever a new version comes out it'll be cleaner, but for today it'll almost certainly break things.

The installation method for the whole package will now just be `pip3 install mmpm --user` followed by `mmpm install --gui`. It sets up all the daemons automatically.

**Comment by khassel** [comment]
````
root@4a3831504b18:~# mmpm install --help
usage:
  mmpm install <package(s)> [--yes]
  mmpm install [--magicmirror] [--autocomplete] [--gui] [--as-module]

optional arguments:
  -h, --help      show this help message and exit
  -y, --yes       assume yes for user response and do not show prompt
  --magicmirror   install MagicMirror, if not already installed
  --autocomplete  install autocompletion for the MMPM CLI
  --gui           install the MMPM GUI. Asks for sudo permissions
  --as-module     install the MMPM MagicMirror helper module in your
                  MagicMirror modules directory to enable hide/show
                  functionality
````

I need `mmpm install --gui -y` or the `make` target you mentioned.

**Comment by Bee-Mar** [comment]
If the root user on your docker image doesn’t need a password to install you could just pipe the yes command.

yes | mmpm install —gui

I intentionally left out the yes option as a possibility because I wanted to user to know it’ll use sudo.

I’ve used the piping above in my own testing on a raspberry pi 3 and 4.

**Comment by Bee-Mar** [comment]
If that still isn’t exactly what you need, then I can set up a make target that’ll do what’s needed.

**Comment by khassel** [comment]
````
root@522e657369e9:~# mmpm install --gui
+ Initializing MagicMirror 3rd party packages database ✓
Checking MMPM [application] (automated) for updates ✓
Are you sure you want to install the MMPM GUI? This requires sudo permission. [yes/y] or [no/n]: y
+ Force removing NGINX and SystemD configs ✓
+ Reloading SystemdD daemon ✘
ERROR: Failed to reload SystemdD daemon. See `mmpm log` for details
+ Restarting NGINX SystemD service ✘
ERROR: Failed to restart NGINX SystemdD daemon. See `mmpm log` for details
MMPM GUI Removed!
+ Copying NGINX and SystemdD service configs + Cleaning confiuration files and resetting SystemdD daemons
✓
+ Reloading SystemdD daemon ✘
ERROR: Failed to reload SystemdD daemon. See `mmpm log` for details
+ Enabling MMPM SystemdD daemon ✘
ERROR: Failed to enable MMPM SystemD service. See `mmpm log` for details
Remove the MMPM GUI? [yes/y] or [no/n]: n
````

next problems, I think we need another procedure ...

**Comment by khassel** [comment]
This was my gui build before:

````
FROM node:buster as builder

ENV HOME=/home/node

RUN set -e; \
    apt-get update; \
    apt-get upgrade -y; \
    apt-get install -y nginx-full git nano; \
    echo 'node ALL=(ALL) NOPASSWD: ALL' >> /etc/sudoers; \
    echo '#!/bin/bash' > /usr/bin/sudo; \
    echo 'exec "$@"' >> /usr/bin/sudo; \
    chmod +x /usr/bin/sudo; \
    chown -R root:root /home/node;

WORKDIR /home/node

ARG MMPM_VERSION
RUN set -e; \
    mkdir -p /home/node/.config/mmpm/log; \
    git clone -b ${MMPM_VERSION} --single-branch https://github.com/Bee-Mar/mmpm.git; \
    cd mmpm; \
    git log -1; \
    sed -i 's|sudo systemctl|#sudo systemctl|g' scripts/make/gui/dependencies; \
    sed -i 's|sudo systemctl|#sudo systemctl|g' scripts/make/daemons/install; \
    make dependencies-gui daemons build-gui-from-src install-gui-from-src;
````

no python required and systemctl commands paralysed ...

**Comment by Bee-Mar** [comment]
Yeah looks like it. The ‘install —gui’ subcommand tries time start an NGINX daemon, which isn’t going to work in a docker image. 


Do you have discord? If so, you can email me your username and ID so we can discuss it easier.

**Comment by Bee-Mar** [comment]
K, I’ll be able to work on a make target tomorrow morning.

I’ll name it ‘docker’ so you’ll be able to call ‘make docker’ after cloning.

**Comment by khassel** [comment]
Thanks, will wait for this, it's already late, local time 23:49 ...

**Comment by khassel** [comment]
@Bee-Mar don't investigate in this, I think I found a solution ... coming back later.

**Comment by Bee-Mar** [comment]
@khassel okay. all the static files and systemd files are located in the package after doing a `pip install`, and assuming it installs in `/usr/local/lib/` you could find the static files and systemd files with something like:

```sh
pip3 install mmpm
VERSION=$(python3 -c "import sys; print(f'{sys.version_info.major}.{sys.version_info.minor}')")
ls /usr/local/lib/python$VERSION/dist-packages/mmpm/
```

you would need the files in `static`, `templates`, and `etc`.

`static` and `templates` would need to be extracted to `/var/www/mmpm`, and the entire `etc` directory could be copied to `/etc`

**Comment by khassel** [comment]
The `latest` images are now containing mmpm version 2.01.

**Comment by Bee-Mar** [comment]
awesome. I'm going to go ahead and close this as being an issue on the repo as well
